Adding mirror scripts to data flow repo

Saif-Marei opened the pull request:
Improved mirror scripts and added documentation.
Combined NSSC and BAS scripts into one "mirror" script.

Refactored globus script to:

- use dictionaries to store rawacf file information (hash, year, month, day, yearmonth)

- only update yyyymm's in master.hashes that were modified in the current script run

tjk584 reviewed (commented):
Some ideas for things to improve/revise:

- [ ] Review python coding practices
    - Change strings to fstrings
    - Remove support for python 2
    - Use argparse library for command line arguments 
- [ ] Separate script and gatekeeper class into two separate files
- [ ] Add documentation summaries to the top of all scripts
- [ ] Update repo readme
- [ ] Review usage of emails - reduce usage where emails are unhelpful
- [ ] Look into using inotify to trigger scripts, or running gatekeeper globus right after sync_ scripts run

Saif-Marei commented:
Made the following additional improvements:

- Exit script right away if no files in holding directory
- Added check to make sure "succeeded" files made it to the mirror
- Moved creation of logger to constructor of Gatekeeper class to make the logger accessible to all Gatekeeper methods
- Made moving blocked, non-matching, and failed files to subdirectory of holding_dir into a Gatekeeper method
- Made a Gatekeeper method to optionally log/email a message and exit the script

tjk584 commented on mirror/download_vt_data:
```diff
@@ -0,0 +1,464 @@
+# What hash program are we using?
+HASHPROG=/usr/bin/sha1sum
+# Date, time and other stuff
+STARTTIME=$(date +%s)
```
Some of these variables are unused - this could be cleaned up. The comments could be trimmed up too.

Saif-Marei replied:
Got rid of unused variables and cleaned up the documentation

tjk584 commented on mirror/batch_bas_sync:
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
```
This file is very similar to the `_weekly` file - they could be combined into one, where the script operation is refined through command line options.

Saif-Marei replied:
Combined weekly and monthly BAS and NSSC scripts into 1 master script where the mirror (BAS or NSSC) and the mode (weekly or monthly) are passed in as arguments.

tjk584 commented on mirror/batch_nssc_sync:
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
```
Same as above, this one could be combined with the `_weekly` script.

tjk584 reviewed (commented):
The additions all look nice, the code is coming along quite nicely. Moving the gatekeeper class to its own script really improved the readability of `gatekeeper_globus.py`.

Some more improvements that could be made:
- [x] Combine the bas/nssc scripts that are copies into a single script. The `batch_*_sync` and `sync_bas/nssc_data.sh` scripts are both very similar, and could be combined to single scripts. The unique variables for each script could be hard-coded into the script, and selected by specifying via the command line arguments
- [x] The bash scripts can be cleaned up a bit. Some ideas:
   - Use `exec &>> $LOGFILE` to send messages to the log file, instead of appending each individual command 
   - Remove any unused variables
   - Clean up the comments (there are too many unhelpful comments in sections)

There's a few other small changes I indicated in the code, but otherwise its looking good so far!

Saif-Marei commented on mirror/gatekeeper_class.py:
```diff
@@ -215,6 +215,30 @@ def send_email(self, message=None, subject=None, from_address=None, recipients=N
         s.sendmail(from_address, recipients, email.as_string())
         s.quit()
 
+    def log_email_exit(self, loglevel, email_flag, exit_flag, msg='', sub=''):
```
Added usage of this function in all applicable instances.

Saif-Marei commented on mirror/gatekeeper_globus.py:
```diff
@@ -1688,5 +626,4 @@ def main():
 
 
-# TODO: check if file exists in the hashes file but not on the mirror
 # TODO: go through all "print" statements and either remove them or add them to logger
```
This TODO comment has been removed as it has been completed

tjk584 commented on mirror/sync_mirror_data.sh:
```diff
@@ -202,19 +182,19 @@ cd "${MIRRORHASHDIR}" || exit
 echo "exit" >> "${SFTPBATCH}"
 # Execute sftp batch file, logging errors to sftp error file
-${SFTP} -b "${SFTPBATCH}" ${USER}@${REMOTEHOST} >> "${LOGFILE}" 2> "${SFTPERRORS}"
+${SFTP} -b "${SFTPBATCH}" ${USER}@${REMOTEHOST} 2> "${SFTPERRORS}"
```
This might prevent stderr from going to the log file - you could try using the `tee` command to send stderr to both the log file and the sftperrors file

Saif-Marei replied:
I think this is OK since I want stderr to only go to the sftperrors file in this case.

Saif-Marei replied:
If any sftp errors occur, they are written to the log file at the end of the script too.

tjk584 reviewed (approved):
Looks good - just a few code syntax and organization suggestions. Just needs to be tested in the data flow now.

Saif-Marei commented on mirror/batch_sync_mirror:
```diff
@@ -0,0 +1,92 @@
+#   Compare our mirror to the BAS/NSSC mirror 1 month at a time going back to January 2006
+#   Download any rawacf files that we are missing. Runs monthly.
+
+MIRROR=$1
```
This is helpful. I added this method to sync_mirror_data.sh, batch_sync_mirror, and download_vt_data (all bash scripts).

Saif-Marei commented on mirror/batch_sync_mirror:
```diff
@@ -0,0 +1,92 @@
+
+elif [[ "${MODE}" == "monthly" ]]
+then
+  # Check if mirror was provided
```
Added argument check to top of script.

Saif-Marei commented on mirror/batch_sync_mirror:
```diff
@@ -0,0 +1,92 @@
+elif [[ "${MODE}" == "monthly" ]]
+then
+  # Check if mirror was provided
+  if [[ "${MIRROR}" == "BAS" ]]
```
Added this usage in both batch_sync_mirror and sync_mirror_data.sh

Saif-Marei commented on mirror/sync_mirror_data.sh:
```diff
@@ -32,96 +28,81 @@
 mkdir -p "${LOCALHASHDIR}"
 mkdir -p "${MIRRORHASHDIR}"
 mkdir -p "${LOCALBLDIR}"
 mkdir -p "${LOCALFAILEDDIR}"
```
Added this change, still need to test it.
